tox.ini

```diff
@@ -11,30 +11,35 @@ deps =
     portend
     pytest-services
     munch
+    {[testenv:typing]deps}
+    {[testenv:lint]deps}
 commands =
-    coverage run --source twine -m pytest {posargs:tests}
+    coverage run --source twine -m pytest tests
     coverage report -m
+    coverage html
+    {[testenv:typing]commands}
```

Comment:
So now in addition to installing these in the `typing` virtualenv we're going to install them in every one of the Python versions too? That's a lot of duplicated usage here. What's the value in doing this and removing `lint` from the `envlist` above? The envlist above is for people to run `tox` and get all of the things we _expect_ them to run.

This also means that _every_ python job on travis will fail independently for (probably) the same reasons. Which will look a lot more intimidating than just 1 job failing in Travis to any new contributor.

Reply:
> The envlist above is for people to run `tox` and get all of the things we _expect_ them to run.

I think that's still true with these changes, which includes both the `typing` and `lint` steps in `testenv`, but results in a more succinct `envlist`, that makes for a simpler CI configuration.

One other benefit for contributors: it enables [`tox --devenv venv`](https://tox.readthedocs.io/en/latest/example/devenv.html) to create a complete virtual environment for running the tools ad-hoc, with different options, in a shell or via editor integration. We don't currently have a documented way to do that.

> So now in addition to installing these in the `typing` virtualenv we're going to install them in every one of the Python versions too? That's a lot of duplicated usage here.

True, though I think it's possible that the `typing` and `lint` environments won't be used as much, since `testenv` has everything.

> This also means that _every_ python job on travis will fail independently for (probably) the same reasons. Which will look a lot more intimidating than just 1 job failing in Travis to any new contributor.

I feel like that's already the case with the test suite, though probably less likely to occur. I grant that seeing the same linting error three times could be irritating/intimidating.

It sounds like we're trying to balance ease for maintainers vs. ease for contributors, where both of those depend on individual experience and opinions. These strikes me as a reasonable compromise, but I'm not married to it.

Reply:
Also, I kept the docs changes minimal, until I knew that this change would be okay. If it is, I'm happy to make more changes to `contributing.rst` in this PR.

Reply:
So, in my opinion, barring the ridiculous overhead of tox creating N virtual environments, running `tox` should be fast for everyone (maintainers, contributors, and anyone else who cares to do so). Now let's think through this, running the tests on 3 versions of Python is good time spent and worth the extra time it takes for `tox` to complete, but is running `typing` and `lint` on *every version of Python* (in addition to installing those for every version) worth the contributors time, or is it good enough to run those once and install them once? Additionally, if someone's trying to iterate quickly, they'll likely want to _just_ run the tests (likely against one or more versions of Python) without caring about linting until they're about ready to create a PR. Trying to do `tox -e py..` to get the tests to run and then having to wait for linting to finish is going to be frustrating and take forever.

I grant you that recent additions of integration tests already make our test suite insufferably slow, but I don't think that's a good reason to make them slower.

Reply:
> I grant you that recent additions of integration tests already make our test suite insufferably slow, but I don't think that's a good reason to make them slower.

As noted in https://github.com/pypa/twine/issues/543#issuecomment-616122506, I've been avoiding this via `PYTEST_ADDOPTS`.

> Trying to do tox -e py.. to get the tests to run and then having to wait for linting to finish is going to be frustrating and take forever.

To quantify this, here's some representative timing from my machine:

|`[testenv]commands`|`time tox -e py38`|
|---|---|
| Just `coverage run -m pytest` | 0m7.104s |
| Plus `coverage` reports | 0m7.597s |
| Plus `typing` |  0m9.537s |
| Plus `lint` | 0m11.319s |

> Additionally, if someone's trying to iterate quickly, they'll likely want to just run the tests

FWIW, I do this by activating a virtual environment and running `pytest` directly, without coverage. Alternatively, as I proposed in "open questions", we could have a separate `[testenv:test]` that defines `deps` and `commands` similar to `[testenv:typing]`, so that iterating becomes `tox -e test`.

---

Again, not married to this, but I find it to be an interesting alternative to common uses of tox, which addresses some small pain points that I've seen.

Reply:
`<bikeshed><yakshave>`

Out of curiosity, I went ahead and extract `[testenv:test]` in https://github.com/pypa/twine/pull/614/commits/2d2f64b9d3955141a1b3a9ddbee5a5143a000259.

`time tox -e test`: 0m5.753s

So, looks like running `coverage` takes about as much time as running `typing`.

`</yakshave></bikeshed>`

Reply:
I'm honestly confused _why_ we're moving away from what has been a fairly easy to follow standard wherein linting is a separate env and (only) tests are run in the `py*` envs. I grant you I haven't been as involved with the larger python community for a while, but we seem to be consistently moving in the wrong direction with this PR to the best of my knowledge.

We haven't explained what are the real benefits of shuffling things around this way. What is this team getting out of this? Also `tox -e test` runs against which version of python? How can I force it to run against a different version?